File: src/test/java/no/novari/configuration/validation/ValidationErrorsFormattingServiceTest.java
package no.novari.configuration.validation;

import no.novari.configuration.validation.ValidationErrorsFormattingService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Path;
import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ValidationErrorsFormattingServiceTest {

    private ValidationErrorsFormattingService service;

    @BeforeEach
    void setUp() {
        service = new ValidationErrorsFormattingService();
    }

    @Test
    void formatNoErrors() {
        Set<ConstraintViolation<Object>> errors = new HashSet<>();
        String result = service.format(errors);
        assertEquals("Validation error: []", result);
    }

    @Test
    void formatSingleError() {
        Set<ConstraintViolation<Object>> errors = new HashSet<>();

        ConstraintViolation<Object> mockViolation = mock(ConstraintViolation.class);
        when(mockViolation.getMessage()).thenReturn("Invalid value");
        Path mockPath = mock(Path.class);
        when(mockPath.toString()).thenReturn("field");
        when(mockViolation.getPropertyPath()).thenReturn(mockPath);

        errors.add(mockViolation);

        String result = service.format(errors);
        assertEquals("Validation error: ['field Invalid value']", result);
    }

    @Test
    void formatMultipleErrors() {
        Set<ConstraintViolation<Object>> errors = new HashSet<>();

        ConstraintViolation<Object> mockViolation1 = mock(ConstraintViolation.class);
        when(mockViolation1.getMessage()).thenReturn("Invalid value");
        Path mockPath1 = mock(Path.class);
        when(mockPath1.toString()).thenReturn("field");
        when(mockViolation1.getPropertyPath()).thenReturn(mockPath1);

        ConstraintViolation<Object> mockViolation2 = mock(ConstraintViolation.class);
        when(mockViolation2.getMessage()).thenReturn("Invalid format");
        Path mockPath2 = mock(Path.class);
        when(mockPath2.toString()).thenReturn("otherField");
        when(mockViolation2.getPropertyPath()).thenReturn(mockPath2);

        errors.add(mockViolation1);
        errors.add(mockViolation2);

        String result = service.format(errors);
        assertEquals("Validation errors: ['field Invalid value', 'otherField Invalid format']", result);
    }

    @Test
    void formatBlankPropertyPath() {
        Set<ConstraintViolation<Object>> errors = new HashSet<>();

        ConstraintViolation<Object> mockViolation = mock(ConstraintViolation.class);
        when(mockViolation.getMessage()).thenReturn("Invalid value");
        Path mockPath = mock(Path.class);
        when(mockPath.toString()).thenReturn("");
        when(mockViolation.getPropertyPath()).thenReturn(mockPath);

        errors.add(mockViolation);

        String result = service.format(errors);
        assertEquals("Validation error: ['Invalid value']", result);
    }
}

File: src/test/kotlin/no/novari/flyt/configuration/validation/ValidationErrorsFormattingServiceTest.kt
package no.novari.flyt.configuration.validation

import jakarta.validation.ConstraintViolation
import jakarta.validation.Path
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class ValidationErrorsFormattingServiceTest {
    private lateinit var service: ValidationErrorsFormattingService

    @BeforeEach
    fun setUp() {
        service = ValidationErrorsFormattingService()
    }

    @Test
    fun formatNoErrors() {
        val errors = mutableSetOf<ConstraintViolation<Any>>()
        val result = service.format(errors)
        assertEquals("Validation error: []", result)
    }

    @Test
    fun formatSingleError() {
        val errors = mutableSetOf<ConstraintViolation<Any>>()

        val mockViolation = mock<ConstraintViolation<Any>>()
        whenever(mockViolation.message).thenReturn("Invalid value")
        val mockPath = mock<Path>()
        whenever(mockPath.toString()).thenReturn("field")
        whenever(mockViolation.propertyPath).thenReturn(mockPath)

        errors.add(mockViolation)

        val result = service.format(errors)
        assertEquals("Validation error: ['field Invalid value']", result)
    }

    @Test
    fun formatMultipleErrors() {
        val errors = mutableSetOf<ConstraintViolation<Any>>()

        val mockViolation1 = mock<ConstraintViolation<Any>>()
        whenever(mockViolation1.message).thenReturn("Invalid value")
        val mockPath1 = mock<Path>()
        whenever(mockPath1.toString()).thenReturn("field")
        whenever(mockViolation1.propertyPath).thenReturn(mockPath1)

        val mockViolation2 = mock<ConstraintViolation<Any>>()
        whenever(mockViolation2.message).thenReturn("Invalid format")
        val mockPath2 = mock<Path>()
        whenever(mockPath2.toString()).thenReturn("otherField")
        whenever(mockViolation2.propertyPath).thenReturn(mockPath2)

        errors.add(mockViolation1)
        errors.add(mockViolation2)

        val result = service.format(errors)
        assertEquals("Validation errors: ['field Invalid value', 'otherField Invalid format']", result)
    }

    @Test
    fun formatBlankPropertyPath() {
        val errors = mutableSetOf<ConstraintViolation<Any>>()

        val mockViolation = mock<ConstraintViolation<Any>>()
        whenever(mockViolation.message).thenReturn("Invalid value")
        val mockPath = mock<Path>()
        whenever(mockPath.toString()).thenReturn("")
        whenever(mockViolation.propertyPath).thenReturn(mockPath)

        errors.add(mockViolation)

        val result = service.format(errors)
        assertEquals("Validation error: ['Invalid value']", result)
    }
}
